### PR TITLE
Prefer GPU, Link blas privately, position PTX Flag correctly

### PR DESCRIPTION
**Pull request by hmaarrfk**
Closes https://github.com/conda-forge/pytorch-cpu-feedstock/issues/114
Closes https://github.com/conda-forge/openblas-feedstock/issues/134
Closes https://github.com/conda-forge/pytorch-cpu-feedstock/issues/102
Closes https://github.com/conda-forge/pytorch-cpu-feedstock/pull/105

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
xref: https://github.com/conda-forge/openblas-feedstock/issues/134

**Comment by ngam** [comment]
maybe include #114 in this too? Should be minor

**Comment by hmaarrfk** [comment]
has a user complained about incompatibility with cuda architectures yet?

Honestly, I'm not so well versed in the different versions. I once tried to hunt down the different options and it seems that many of them reduced to the same compile flags.

the only thing we seem to be missing is likely "3.7". Lets wait until somebody complain.
https://github.com/pytorch/pytorch/blob/master/cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake#L220

**Comment by ngam** [comment]
No, not that I am aware of. I am also fine with waiting (or potentially dropping the older GPUs as pytorch upstream drops them). 

I am still confused about the validity and usefulness of PTX here. I have a negative feeling about it tbh, because I am afraid it is stalling the compilation because it is pushing to take into accounts things are already taken into account (i.e. 5.0+PTX means 5.0 and all after it, but we are already listing all those after it). I will try to run this separately to see what's going on, but for this PR, we can ignore it :)

**Comment by ngam** [comment]
> Honestly, I'm not so well versed in the different versions. I once tried to hunt down the different options and it seems that many of them reduced to the same compile flags.

I found this useful https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/ but I am not sure if this is what you're talking about

**Review by ngam — APPROVED** [review]
Thanks!

One note: I am not sure how the pytorch pin will work since this only elevates the build number, specifically I am not sure if a rebuild is required, but you're more knowledgable of that stuff

**Comment by hmaarrfk** [comment]
> One note: I am not sure how the [pytorch pin](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/1a62c5855e685402f2ecf858d3cdc51b8c0d9fe4/recipe/conda_build_config.yaml#L712-L713) will work since this only elevates the build number, specifically I am not sure if a rebuild is required, but you're more knowledgable of that stuff

I'm not sure. To be honest, pytorch is somewhat seldom used as a library. I think the way it is packaged now means that people are jumping through hoops they shouldn't need to.

I was assuming I could break things.

Maybe we can repo patch to pin to a specific build. @jakirkham do you know if this is possible?

**Comment by ngam** [comment]
Yes, I have no idea... but I stand by my unsolicited "approve" here and in #105. In my understanding, this introduces solutions and no new problems. So it is a step forward.

**Comment by peastman** [comment]
I strongly advise against this patch.  I do not believe it is the correct fix.  It changes the binary interface to libtorch and will break any downstream package that expects linking to libtorch to bring in BLAS routines.

The real problem this is trying to work around is simply that the conda-forge version of the `openblas` package is missing a single alias that is present in all other BLAS packages, including other versions of the `openblas` package, and that PyTorch expects to be present.

**Comment by ngam** [comment]
> I strongly advise against this patch. I do not believe it is the correct fix. It changes the binary interface to libtorch and will break any downstream package that expects linking to libtorch to bring in BLAS routines.
> 
> The real problem this is trying to work around is simply that the conda-forge version of the `openblas` package is missing a single alias that is present in all other BLAS packages, including other versions of the `openblas` package, and that PyTorch expects to be present.

It would be helpful to open a new issue in this feedstock explaining your objection in terms of the pytorch-feedstock and pytorch, in as much detail as you can. From what I understand from the other issue, your objection is simply that MKL is doing it a certain way, and so should OpenBLAS. That may not be enough, I am afraid.

**Comment by hmaarrfk** [comment]
I'm happy to delay until the 1.12.0 release

**Comment by leofang** [comment]
> will break any downstream package that expects linking to libtorch to bring in BLAS routines.

Why would downstream packages expect this? If they need BLAS, they should link to it explicitly, otherwise there could be issues when dropping in a different BLAS implementation.

**Comment by hmaarrfk** [comment]
> Why would downstream packages expect this

I'm not sure they expect this, so much as likely already linked to it erroneously. I'm not an expert at how compilers work, so I am worried about removing symbols unexpectidly from the pytorch libraries.

**Comment by hmaarrfk** [comment]
[log_files.zip](https://github.com/conda-forge/pytorch-cpu-feedstock/files/8857307/log_files.zip)

I'm uploading the builds to my channel now under the cfep03 label. 
https://anaconda.org/mark.harfouche/pytorch/files

I'm going to leave this up for 24 hours before uploading.

**Comment by hmaarrfk** [comment]
before copying them to conda-forge.

Thank you all for your input. Glad to keep improving the state of this package!

**Comment by ngam** [comment]
Confirmed:

```
~$ mamba create -n test_pt -c mark.harfouche/label/cfep03 pytorch==*=*cuda112*

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.24.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['pytorch===[build=*cuda112*]']

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache
mark.harfouche/label/cfep03/linux-64                          No change
mark.harfouche/label/cfep03/noarch                            No change

Encountered problems while solving:
  - nothing provides __cuda needed by pytorch-1.11.0-cuda112py38habe9d5a_202

```

**Comment by hmaarrfk** [comment]
A slightly modified patch was accepted upstream. The upstreamed patch included a few changes for the tests that we don't use.

Truthfully, I don't understand why we didn't need the changes for pybind11
https://github.com/pytorch/pytorch/pull/78883/files#diff-c5ee05f1e918772792ff6f2a3f579fc2f182e57b1709fd786ef6dc711fd68b27R2011


However, if you don't think this is an issue, i'm glad to merge as is
Please find logs attached. [log_files.zip](https://github.com/conda-forge/pytorch-cpu-feedstock/files/8874416/log_files.zip)

**Comment by hmaarrfk** [comment]
The reason why we didn't need the pybind part of the patch is likely because MKLDNN was never found. This is true also in the PRs in the main branch:

```
2022-04-27T01:46:32.2129230Z   -- MIOpen not found. Compiling without MIOpen support
2022-04-27T01:46:32.2880311Z   -- MKLDNN source files not found!
2022-04-27T01:46:32.2887271Z   CMake Warning at cmake/Dependencies.cmake:1754 (message):
2022-04-27T01:46:32.2894893Z     MKLDNN could not be found.
2022-04-27T01:46:32.2901842Z   Call Stack (most recent call first):
2022-04-27T01:46:32.2907922Z     CMakeLists.txt:669 (include)
2022-04-27T01:46:32.2913734Z 
```

**Comment by hmaarrfk** [comment]
Files are located: https://anaconda.org/mark.harfouche/pytorch/files

**Comment by hmaarrfk** [comment]
thank you all for your input. CPU builds should be available soon.

**Comment by ngam** [comment]
> The reason why we didn't need the pybind part of the patch is likely because MKLDNN was never found. This is true also in the PRs in the main branch:
> 
> ```
> 2022-04-27T01:46:32.2129230Z   -- MIOpen not found. Compiling without MIOpen support
> 2022-04-27T01:46:32.2880311Z   -- MKLDNN source files not found!
> 2022-04-27T01:46:32.2887271Z   CMake Warning at cmake/Dependencies.cmake:1754 (message):
> 2022-04-27T01:46:32.2894893Z     MKLDNN could not be found.
> 2022-04-27T01:46:32.2901842Z   Call Stack (most recent call first):
> 2022-04-27T01:46:32.2907922Z     CMakeLists.txt:669 (include)
> 2022-04-27T01:46:32.2913734Z 
> ```

hmmm

I don't know, but I thought the conda-forge's build did find mkl and mkldnn successfully (not that I care). I will try to have another look later on my end locally. Later in the summer, I am rewriting some tensorflow pipelines in pytorch and jax, so I will see how things end up --- I will try to communicate my findings, like I did in the tensorflow feedstock... hopefully we can get conda-forge's pytorch to beat the NGC pytorch containers :)

**Comment by ngam** [comment]
> thank you all for your input. CPU builds should be available soon.

Thank YOU 🌟

**Comment by hmaarrfk** [comment]
> I don't know, but I thought the conda-forge's build did find mkl and mkldnn successfully (not that I care). I will try to have another look later on my end locally. Later in the summer, I am rewriting some tensorflow pipelines in pytorch and jax, so I will see how things end up --- I will try to communicate my findings, like I did in the tensorflow feedstock... hopefully we can get conda-forge's pytorch to beat the NGC pytorch containers :)

That would be cool.
